docs: Refreshing SSH documentation

garethtdavies opened the pull request:
Closes #1263 
Closes #1238 
Closes #1234 
Closes #1147 

Change-type: patch
Signed-off-by: Gareth Davies gareth@balena.io

garethtdavies commented:
I've refreshed these docs to add the required SSH key info.

I’ve no idea if this section (Install an SSH server in a container) is still relevant https://github.com/balena-io/docs/pull/1239/files#diff-e7009d4cdf42562438e902abe2e20aabR49 perhaps @shaunmulligan can comment. There’s a link to a gitter discussion so I imagine a lot of this is very dated.

garethtdavies commented:
There is also this issue: https://github.com/balena-io/docs/issues/1116 specifically:

`Talk through adding sshKeys field to config.json for ssh in production devices.`

I'm documenting `config.json` elsewhere in a separate PR but to confirm if I add an SSH key via the control panel this also works for a production image?

pdcastro commented:
That page is only for `balena ssh`, right? I wonder if any of the following forum comments would be relevant here:
* https://forums.balena.io/t/rsync-over-balena-ssh-tunnel/6228/4
* https://forums.balena.io/t/howto-ssh-into-host-device/5312

garethtdavies commented:
Thanks for those links @pdcastro. I've pushed an update based on this and some additional testing.

I've updated this PR with standalone SSH commands, and I've also included details of tunneling the connection if the port is blocked (from your forum post).

I think the rsync/scp material makes for a great article in a "how to section" (that doesn't yet exist :) ) that references these docs.

I'm not clear on the second link as this seems more related to openBalena? Everything seems to be possible just via a combination of balenaCLI, standalone SSH, balena tunnel, and updating config.json so I'm not sure what else in this article would be required?

Anyway, I've updated this PR and would love more feedback as it already touches a _lot_ of open issues so clearly a point of some confusion. There is an intentionally broken link in this doc to updating config.json as this page doesn't exist, but there is an open PR for that here https://github.com/balena-io/docs/pull/1249.

Finally, I've removed the "Install an SSH server in a container" section, as I think this is redundant as everything seems possible to do with all the approaches already detailed in this PR. If I've misunderstood this, then someone, please correct me!

pdcastro commented on shared/general/container-ssh.md:
```diff
@@ -1,46 +1,83 @@
+
+This also works in multicontainer applications, simply pass the name of the appropriate service as defined in docker-compose.yml you'd like to access the shell for.
+
+For devices running development images on your local network, you can create a direct SSH connection by specifying the hostname using the `<device-uuid>.local` syntax or local IP address of the device. Development images have passwordless root SSH access enabled by default, and so they should never be directly exposed to the public internet. Using {{ $names.company.short }} ssh in this way does n ...
```
> For devices running development images on your local network, you can create a direct SSH connection

I found this a bit confusing, because a "direct SSH connection" (meaning a connection that does not rely on the balenaCloud VPN) can be used with both development images and production images, and it is also not limited to local IP addresses -- a globally routable IP address would work too, although uncommon. I suggest (just a suggestion) reordering and slightly rewording this paragraph and the following two, as follows:

---
> ...
> When an application name or device UUID is used as above, `{{ $names.company.short }} ssh` uses the {{ $names.company.short }} VPN to create a secure tunnel to the device and then forward SSH traffic between the device and your development machine. For this method, you must have a SSH key configured on your development machine and [added to the {{ $names.cloud.lower }} dashboard][add-ssh-key].
>
> If an IP address or a `.local` hostname is used (instead of an application name or device UUID), `{{ $names.company.short }} ssh` establishes a direct connection that does not rely on the {{ $names.company.short }} VPN:
>
```shell
$ balena ssh 192.168.1.23
$ balena ssh <device-uuid>.local
```
>
> This should work without further configuration in the case of devices running a [development balenaOS image](https://www.balena.io/docs/reference/OS/overview/2.x/#dev-vs-prod-images), which allows passwordless root SSH access (and for this reason, should never be directly exposed to the public internet). In the case of a production balenaOS image, an SSH key needs to be present in the `config.json` file of the device. More details about configuring SSH keys in `config.json` may be found [here][config-json-ssh].

Above I have also replaced "development image" with "development balenaOS image", and made it a link.

garethtdavies replied:
Thank you, this is excellent feedback and I have updated this.

garethtdavies commented on shared/general/container-ssh.md:
```diff
@@ -1,46 +1,83 @@
 
-<img src="/img/common/device/terminal.png" width="60%">
-
-In order to start a terminal session for a service, you first need to ensure that your device is online and that the service container is running. If the container code crashes or ends quickly, it is not possible to attach a console to it. Check out our [Dockerfile][dockerfile-init] and [Docker Compose][compose-init] guides for more information on keeping your containers running.
```
Right, this was an over-aggressive purge of outdated init links! Reworded this and avoided another note :) I actually had this issue testing this as many of our hello world projects simply print "hello world" and exit rather than looping/sleeping.

garethtdavies commented:
@balena-ci retest
